pgrx/src/datum/datetime_support/chrono.rs

```diff
@@ -0,0 +1,194 @@
+//!
+//! Note that `chrono` has no reasonable analog for the `time with timezone` (i.e. [`pgrx::TimeWithTimeZone`]), so there are no added conversions for that type outside of the ones already implemented.
+#![cfg(feature = "chrono")]
+
+use core::convert::Infallible;
+use core::num::TryFromIntError;
+use std::convert::TryFrom;
+
+use chrono::{DateTime, Datelike, NaiveDate, NaiveDateTime, NaiveTime, Timelike, Utc};
+
+use crate::datum::datetime_support::DateTimeConversionError;
+use crate::datum::{Date, Time, Timestamp, TimestampWithTimeZone};
+
+/// Convenience type for [`Result`]s that fail with a [`DateTimeConversionError`]
+type DtcResult<T> = Result<T, DateTimeConversionError>;
+
+impl From<TryFromIntError> for DateTimeConversionError {
+    fn from(_tfie: TryFromIntError) -> Self {
+        DateTimeConversionError::FieldOverflow
+    }
+}
+
+impl From<Infallible> for DateTimeConversionError {
+    fn from(_i: Infallible) -> Self {
+        DateTimeConversionError::FieldOverflow
+    }
+}
+
+impl TryFrom<Date> for NaiveDate {
+    type Error = DateTimeConversionError;
+
+    fn try_from(d: Date) -> DtcResult<NaiveDate> {
+        NaiveDate::from_ymd_opt(d.year(), d.month().into(), d.day().into())
+            .ok_or_else(|| DateTimeConversionError::InvalidFormat)
+    }
+}
+
+impl TryFrom<NaiveDate> for Date {
+    type Error = DateTimeConversionError;
+
+    fn try_from(d: NaiveDate) -> DtcResult<Date> {
+        let month = u8::try_from(d.month())?;
+        let day = u8::try_from(d.day())?;
+        Date::new(d.year(), month, day)
+    }
+}
+
+/// Note: conversions from Postgres' `time` type [`pgrx::Time`] to [`chrono::NaiveTime`]
+/// incur a loss of precision as Postgres only exposes microseconds.
+impl TryFrom<Time> for NaiveTime {
+    type Error = DateTimeConversionError;
+
+    fn try_from(t: Time) -> DtcResult<NaiveTime> {
+        let (hour, minute, second, microseconds) = t.to_hms_micro();
+        let seconds_micro: u32 = Into::<u32>::into(second)
+            .checked_mul(1_000_000)
+            .ok_or(DateTimeConversionError::FieldOverflow)?;
+        NaiveTime::from_hms_micro_opt(
+            hour.into(),
+            minute.into(),
+            second.into(),
+            // Since pgrx counts the fractional seconds (between 1_000_000 and 2_000_000),
+            // the microseconds value will be 1_000_000 * seconds + fractional.
+            //
+            // - at 12:01:01 => hour=12, minute=1, second=2, microseconds=2_000_000
+            // - at 12:01:59 => hour=12, minute=1, second=59, microseconds=59_000_000
+            // - at 12:01:59 => hour=12, minute=1, second=59, microseconds=59_000_000
+            //
+            // Since chrono *does* support leap seconds (representing them as 59 seconds & >1_000_000 microseconds)
+            // we can strip the microseconds in that case to zero and pretend they're not there,
+            // since Postgres does not support leap seconds
```

Comment:
Thank you for documenting these little decisions.

Reply:
Absolutely -- did this mostly so that people would at least have a better chance of finding my bugs if there are some hiding in here.

Time is one of the places where programming goes off the rails, I'm just *hoping* I mostly got it right.